### PR TITLE
[core] Simplify util::peer

### PR DESCRIPTION
**Pull request by pozdnyakov**

### CODE REVIEW

**Comment by kkaefer** [comment]
This changes the data from inline/stack to heap. Is that intentional?

File: include/mbgl/util/peer.hpp [hunk 1]
```diff
@@ -8,103 +9,30 @@ namespace util {
...
-
-    template <class T>
-    static vtable* get_vtable() {
-        static vtable_impl<T> vtable;
-        return &vtable;
-    }
-
-    vtable* vtable = nullptr;
-    storage_t storage;
+    std::unique_ptr<DataHolderBase> ptr;
```
Comment (kkaefer) [inline include/mbgl/util/peer.hpp]: This changes from inline allocation to heap allocation. Is that intended?

Comment (pozdnyakov) [reply include/mbgl/util/peer.hpp]: The intention was to rely on the standard `std::unique_ptr` class and achieve the same functionality for the client with a simpler code. It changes to heap allocation indeed, however peer creation does not seem to happen often enough to make a performance impact.

Comment (kkaefer) [reply include/mbgl/util/peer.hpp]: Yeah, you're probably right. I'm curious though why this change _increases_ the binary size on iOS? any ideas?

Comment (pozdnyakov) [reply include/mbgl/util/peer.hpp]: > I'm curious though why this change _increases_ the binary size on iOS? any ideas?

this is something I'm looking atm, this increase is quite surprising - I expected the opposite :)

Comment (pozdnyakov) [reply include/mbgl/util/peer.hpp]: @kkaefer the peer in the latest patch is wrapping `std::unique_ptr<void, std::function<void(void*)>>` and it limits the binary size increase on iOS quite significantly. Could you take a look?

**Pull request merged**